### ISSUE TITLE
YapDatabaseCloudKit with Swift value types

### ISSUE THREAD

**Issue by heiberg** [yapstudios/YapDatabase#291]
**TLDR:** Best strategy when you can't use KVO and stateful/mutable model objects?

_(I'd tag this as a question, because I'm really just seeking advice here.)_

**Current setup:** I have an app, written in Swift, which uses YapDatabase. My models are plain structs. The objects going in and out of the database are NSCoding compliant helper classes, which are created on-demand when reading and writing. These helpers are short lived and not passed around the app. 

(In case it's relevant, the application in question is using @danthorpe 's delightful [YapDatabaseExtensions](https://github.com/danthorpe/YapDatabaseExtensions) to accomplish this.)

Now I am trying to add CloudKit sync to this app, learning as I go from the CloudKitTodo example.

The approach taken in CloudKitTodo with a smart superclass (MyDatabaseObject) won't work in this environment, so I'm trying to come up with a good alternative to MyDatabaseObject.

This is where I'm feeling a bit stuck, as I'm still pretty new to both YapDatabase and CloudKit.

In order to implement the record handler and merge blocks I need to answer questions about what changed in the object.

By using Swift structs I've opted out of KVO and type introspection. This is obviously a trade-off and I'm perfectly happy to diff two instances myself, field by field.

I think my question is: 

**What would be an elegant way to get access to both the original and modified value in the record handler block?**

I believe the recordHandler is invoked too late for me to read the original from the database?

I could use my NSCoding helper to add a shadow copy to my struct. This would carry over unchanged when creating new, modified values and be available when saving. Conceptually something like this (ignore that recursive value types are illegal):

```swift
struct Person {
	let name: String
	let syncInfo: Person? // Not persisted. Only read when saving.
}
```

Or perhaps I could duplicate the data in the database. Either in a parallel collection or in the metadata. This would double my local storage cost, but my data is so small that it would be acceptable.

Maybe I could somehow compare against fields in the CKRecord? If no CKRecord is found it's a newly inserted object and all fields must be updated.

I apologize if this comes of as a bit rambling. I feel like I just barely know enough to phrase the question.

**Comment by danthorpe** [comment]
Hi @heiberg - I think this is a great question. And I have scratched my head too while working on exactly the same problem. Unfortunately I've not yet had the time to make any kind of open source solution to this sort of problem (Value Type Models + Database + CloudKit).

Getting to your question - because that is really the heart of the matter, I think you are basically on the right track, except that I would put the previous/modified value (or properties of the value type?) into your `CoderType` (speaking in [ValueCoding](https://github.com/danthorpe/ValueCoding) terms) rather than the value type itself.

For a slightly longer suggestion: I'm assuming that you'll have already defined various protocols/types to make syncing your types "easy" (or at least straightforward). I can't really share here how I did this myself, but essentially I had a `Syncable` protocol which exposed via a static variable a "sync info" object which contained metadata about the properties which should be synced (again, using general types). Using this, it should be fairly easy to create a base `CoderType` which accepts a generic `Syncable` and supports keeping track of the modified properties.

In your cloud kit record handler, you should be receiving your archives (as YapDatabase knows nothing of your value types) so you should be able to check for changed properties.

Hopefully that gives you some ideas.

I'd be really interested in how you get on, as I've done some initial work on making helpers for this, but I've got side tracked on other things recently.

**Comment by heiberg** [comment]
Thanks a lot, @danthorpe! I think that put me on a good path.

If I end up with a reusable set of sync helpers I'll be happy to contribute them.

**Comment by heiberg** [comment]
In case someone finds this, and because @danthorpe expressed interest in an update, [I put up this description of how I ended up solving this.](https://gist.github.com/heiberg/5f596aafe5820f767dbe)

There's a file with some reusable helper code there too.

**Comment by danthorpe** [comment]
This is great stuff @heiberg! Will comment more in the gist. :)
